### ISSUE TITLE
Q: HMMRATAC producing too many peaks

### ISSUE THREAD

**Issue by igordot** [macs3-project/MACS#636]
I ran HMMRATAC on two samples. The quality difference between them is high, although the quality is far from ideal for both. However, the low quality sample produced more than 10x as many peaks (<10k versus >100k) with the default settings.

For comparison, the Java version of HMMRATAC produced <1k peaks for the first sample and no peaks for the second sample (score cutoff of 30).

Here is a genome browser screenshot to illustrate the situation:
<img width="939" alt="image" src="https://github.com/macs3-project/MACS/assets/6363505/5ad480b5-7b4d-4991-970f-931602786840">

Is this expected? Are there some guidelines to help avoid this?

**Comment by taoliu** [comment]
One thing you can try is to increase the cutoff for calling candidate regions in the macs3 version of `hmmratac` using the option for prescan-cutoff `-c`. And perhaps first run the `--cutoff-analysis` in `hmmratac` to decide the OK parameter for `-c`.  Also it is likely that the cutoff values for `-u` and `-l` for training HMM have to be adjusted as well. Check https://macs3-project.github.io/MACS/docs/hmmratac.html#choices-of-cutoff-values

Please let me know if this can help.

**Comment by igordot** [comment]
I saw that, but wasn't sure how to properly use the output to determine the optimal the values. Do you have an example documented somewhere?

**Comment by taoliu** [comment]
The cutoff-analysis for `hmmratac` command is discussed here:  https://macs3-project.github.io/MACS/docs/hmmratac.html#choices-of-cutoff-values

And a similar cutoff-analysis for `callpeak` command is discussed here: https://macs3-project.github.io/MACS/docs/callpeak.html#cutoff-analysis

And if you can generate the cutoff analysis report from `hmmratac --cutoff-analysis-only` and share with me here, we can analyze the report together in this thread.

**Comment by igordot** [comment]
I am sharing the cutoff report here (I had to add .txt because curiously GitHub does not allow .tsv):
[test_cutoff_analysis.tsv.txt](https://github.com/macs3-project/MACS/files/14912573/test_cutoff_analysis.tsv.txt)

**Comment by taoliu** [comment]
The fold change values for reasonable peaks are within the lower range.  The default parameters for `hmmratac` are `-u 20 -l 10 -c 1.2`. It's possible that we need to increase the `-c` parameter a bit higher. I can't judge since the current npeak, lpeak, and avelpeak values of the score cutoff range 0.00-8.40 show huge difference and we don't have a finer resolution of the analysis between 0 and 8.4. You can try to use `-u 20 -l 8 -c 5`, and to see if the result can improve. We will try to add an option for fine-tuning this cutoff analysis to get better resolution in the future. At the meantime, we are adding a different emission model -- Poisson emission for HMM, as an addition to the current Gaussian emission model. In our internal test on some small dataset, it gives us more similar results as the java version HMMRATAC.

How is the current HMM from `hmmratac`? Could you share us the command line output as https://macs3-project.github.io/MACS/docs/hmmratac.html#tune-the-hmm-model?

**Comment by Jungal10** [comment]
> I ran HMMRATAC on two samples that should be biological replicates. The quality difference between them is high, although the quality is far from ideal for both. However, the low quality sample produced more than 10x as many peaks (<10k versus >100k) with the default settings.
> 
> For comparison, the Java version of HMMRATAC produced <1k peaks for the first sample and no peaks for the second sample (score cutoff of 30).
> 
> Here is a genome browser screenshot to illustrate the situation: <img alt="image" width="939" src="https://private-user-images.githubusercontent.com/6363505/319232205-5ad480b5-7b4d-4991-970f-931602786840.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI4NDExMTMsIm5iZiI6MTcxMjg0MDgxMywicGF0aCI6Ii82MzYzNTA1LzMxOTIzMjIwNS01YWQ0ODBiNS03YjRkLTQ5OTEtOTcwZi05MzE2MDI3ODY4NDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTFUMTMwNjUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmJjN2U4YzlhZGUwZTI2OTM5OGE1ZDhmMGJjYzk1MDVjYTMzNzgyN2E1ZGVhODY5NzU4MWRkNTU4NDEzOWZjNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.fd0k65lv5NVOlholD06TH3EmHV6WHCGRepI-NtbNihs">
> 
> The called peaks for the first sample look like peaks. Not so much for the second sample. Is this expected? Are there some guidelines to help avoid this?

Not a direct answer, but the scale on your IGV is quite different as well between the two

**Comment by igordot** [comment]
> Not a direct answer, but the scale on your IGV is quite different as well between the two

Yes, each sample is auto-scaled independently since peak-calling is independent as well.

**Comment by taoliu** [comment]
Also related to #638, it's highly possible that the hidden markov model learned from the low quality data can't capture the expected chromatin structure around accessible sites. This is a common problem of "machine learning". In this case, perhaps a general peak calling process such as using "callpeak" function is more appropriate.

**Comment by igordot** [comment]
I am not surprised that machine learning is failing for these samples. I was primarily concerned about the large difference between the Java and MACS3 implementations.
